File: packages/sdk/src/http.ts
/**
 * Minimal HTTP client wrapping native `fetch`.
 *
 * Handles:
 * - Auth header injection (`X-API-Key`)
 * - JSON request/response serialization
 * - Error mapping to typed SDK errors
 * - Request timeout via AbortController
 * - SSE stream parsing for exec/stream endpoint
 * - Exponential-backoff retry on idempotent operations (GET/DELETE)
 */

import {
  mapApiError,
  SandboxError,
  TimeoutError,
  ValidationError,
} from "./errors.js"
import type { ApiExecStreamEvent } from "./types.js"

const DEFAULT_TIMEOUT_MS = 30_000

const SDK_VERSION = "0.7.7"
const USER_AGENT = `@superserve/sdk/${SDK_VERSION} (node/${
  typeof process !== "undefined" && process.versions?.node
    ? `v${process.versions.node}`
    : "unknown"
})`

// Retry tuning
const DEFAULT_MAX_ATTEMPTS = 3
const BASE_BACKOFF_MS = 100
const MAX_BACKOFF_MS = 30_000
const RETRYABLE_STATUSES = new Set([429, 502, 503, 504])

const DEFAULT_MAX_DOWNLOAD_BYTES = 2 * 1024 * 1024 * 1024 // 2 GiB; matches boxd's server-side zip cap

interface RequestOptions {
  method: "GET" | "POST" | "PATCH" | "DELETE"
  url: string
  headers?: Record<string, string>
  body?: unknown
  timeoutMs?: number
  signal?: AbortSignal
}

/**
 * Compose an internal controller signal with an optional user signal.
 * Uses AbortSignal.any when available.
 */
function composeSignals(
  internal: AbortSignal,
  user?: AbortSignal,
): AbortSignal {
  if (!user) return internal
  return AbortSignal.any([internal, user])
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

/**
 * Parse a Retry-After header value.
 * Returns milliseconds to wait, capped at MAX_BACKOFF_MS, or null if invalid.
 */
function parseRetryAfter(value: string | null): number | null {
  if (!value) return null
  const trimmed = value.trim()
  if (!trimmed) return null
  // Numeric seconds
  const asNumber = Number(trimmed)
  if (Number.isFinite(asNumber) && asNumber >= 0) {
    return Math.min(asNumber * 1000, MAX_BACKOFF_MS)
  }
  // HTTP-date
  const asDate = Date.parse(trimmed)
  if (Number.isFinite(asDate)) {
    const delta = asDate - Date.now()
    if (delta <= 0) return 0
    return Math.min(delta, MAX_BACKOFF_MS)
  }
  return null
}

function backoffDelay(attempt: number): number {
  // Attempt 1 → 100ms, attempt 2 → 200ms, attempt 3 → 400ms, etc.
  const base = BASE_BACKOFF_MS * 2 ** (attempt - 1)
  const jitter = base * (0.8 + Math.random() * 0.4)
  return Math.min(jitter, MAX_BACKOFF_MS)
}

/**
 * Is this error a network-level transient error worth retrying?
 *
 * `fetch` throws a `TypeError` for network failures (connection refused,
 * DNS failure, reset, etc). AbortErrors should NOT be treated as network
 * errors — they're explicit cancellations.
 */
function isNetworkError(err: unknown): boolean {
  if (err instanceof DOMException && err.name === "AbortError") return false
  if (err instanceof TypeError) return true
  // Node fetch sometimes surfaces errors with name "FetchError" or a cause
  if (
    err instanceof Error &&
    (err.name === "FetchError" || err.name === "NetworkError")
  ) {
    return true
  }
  return false
}

/**
 * Internal fetch wrapper with per-attempt timeout, abort-signal composition,
 * and optional retry on transient conditions (429, 5xx, network errors).
 *
 * Only retries when `opts.retryable` is true. Callers must ensure the
 * operation is idempotent before enabling retries.
 */
async function retryableFetch(
  input: RequestInfo | URL,
  init: RequestInit,
  opts: {
    timeoutMs: number
    maxAttempts?: number
    retryable: boolean
    userSignal?: AbortSignal
  },
): Promise<Response> {
  const maxAttempts = opts.retryable
    ? (opts.maxAttempts ?? DEFAULT_MAX_ATTEMPTS)
    : 1

  let lastError: unknown
  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    const controller = new AbortController()
    let timedOut = false
    const timer = setTimeout(() => {
      timedOut = true
      controller.abort()
    }, opts.timeoutMs)

    const signal = composeSignals(controller.signal, opts.userSignal)

    try {
      const res = await fetch(input, { ...init, signal })

      // Retry on specific 5xx / 429
      if (opts.retryable && RETRYABLE_STATUSES.has(res.status)) {
        if (attempt >= maxAttempts) {
          return res
        }
        let delay: number | null = null
        if (res.status === 429) {
          delay = parseRetryAfter(res.headers.get("Retry-After"))
        }
        if (delay === null) {
          delay = backoffDelay(attempt)
        }
        // Drain body so the connection can be released
        try {
          await res.arrayBuffer()
        } catch {
          // ignore
        }
        clearTimeout(timer)
        await sleep(delay)
        continue
      }

      return res
    } catch (err) {
      lastError = err

      // User cancellation propagates immediately, never retried
      if (opts.userSignal?.aborted) {
        throw err
      }

      // Internal timeout — surface as TimeoutError
      if (timedOut) {
        throw new TimeoutError(`Request timed out after ${opts.timeoutMs}ms`)
      }

      // Non-network / non-timeout aborts (rare): treat as cancellation
      if (err instanceof DOMException && err.name === "AbortError") {
        throw err
      }

      // Retry network errors if retryable and attempts remain
      if (opts.retryable && isNetworkError(err) && attempt < maxAttempts) {
        await sleep(backoffDelay(attempt))
        continue
      }

      throw err
    } finally {
      clearTimeout(timer)
    }
  }

  // Shouldn't be reachable — the loop either returns or throws
  throw lastError ?? new SandboxError("Request failed with no response")
}

async function readErrorBody(
  res: Response,
): Promise<{ error?: { code?: string; message?: string } }> {
  try {
    return (await res.json()) as { error?: { code?: string; message?: string } }
  } catch {
    return {}
  }
}

/**
 * Make an HTTP request and return parsed JSON.
 *
 * Throws typed SandboxError subclasses on non-2xx responses.
 *
 * Retries GET/DELETE on transient failures (429, 502/503/504, network errors).
 * POST/PATCH are never retried (not idempotent).
 */
export async function request<T>(opts: RequestOptions): Promise<T> {
  const {
    method,
    url,
    headers = {},
    body,
    timeoutMs = DEFAULT_TIMEOUT_MS,
    signal: userSignal,
  } = opts

  const retryable = method === "GET" || method === "DELETE"
  const mergedHeaders: Record<string, string> = {
    "User-Agent": USER_AGENT,
    "Content-Type": "application/json",
    ...headers,
  }

  try {
    const res = await retryableFetch(
      url,
      {
        method,
        headers: mergedHeaders,
        body: body !== undefined ? JSON.stringify(body) : undefined,
      },
      { timeoutMs, retryable, userSignal },
    )

    if (!res.ok) {
      const errorBody = await readErrorBody(res)
      throw mapApiError(res.status, errorBody)
    }

    // 204 No Content
    if (res.status === 204) {
      return undefined as T
    }

    // Some endpoints legally return 2xx with an empty body.
    const text = await res.text()
    return text ? (JSON.parse(text) as T) : (undefined as T)
  } catch (err) {
    if (err instanceof SandboxError) throw err
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new SandboxError("Request aborted", undefined, undefined, {
        cause: err,
      })
    }
    throw new SandboxError(
      `Network error: ${err instanceof Error ? err.message : String(err)}`,
      undefined,
      undefined,
      { cause: err },
    )
  }
}

/**
 * Make a request that returns no body (204 expected).
 */
export async function requestVoid(opts: RequestOptions): Promise<void> {
  await request<void>(opts)
}

/**
 * Upload raw bytes to a URL. Used for file upload to the data plane.
 *
 * Not retried — POST is not idempotent.
 */
export async function uploadBytes(opts: {
  url: string
  headers: Record<string, string>
  body: BodyInit
  timeoutMs?: number
  signal?: AbortSignal
}): Promise<void> {
  const {
    url,
    headers,
    body,
    timeoutMs = DEFAULT_TIMEOUT_MS,
    signal: userSignal,
  } = opts

  const mergedHeaders: Record<string, string> = {
    "User-Agent": USER_AGENT,
    "Content-Type": "application/octet-stream",
    ...headers,
  }

  try {
    const res = await retryableFetch(
      url,
      { method: "POST", headers: mergedHeaders, body },
      { timeoutMs, retryable: false, userSignal },
    )

    if (!res.ok) {
      const errorBody = await readErrorBody(res)
      throw mapApiError(res.status, errorBody)
    }
  } catch (err) {
    if (err instanceof SandboxError) throw err
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new SandboxError("Upload aborted", undefined, undefined, {
        cause: err,
      })
    }
    throw new SandboxError(
      `Upload error: ${err instanceof Error ? err.message : String(err)}`,
      undefined,
      undefined,
      { cause: err },
    )
  }
}

/**
 * Read a response body into a single Uint8Array, refusing to buffer more than
 * `maxBytes`.
 *
 * The data plane is untrusted: a hostile sandbox can return a multi-GB or
 * endless response. Streaming the body lets us stop pulling bytes (and free the
 * connection) the moment the running total exceeds the cap, instead of
 * buffering the whole thing first.
 */
async function readBodyWithLimit(
  res: Response,
  maxBytes: number,
): Promise<Uint8Array> {
  // No streaming body available (e.g. mocked responses) — fall back to a full
  // buffer read, then enforce the cap.
  if (!res.body) {
    const buf = new Uint8Array(await res.arrayBuffer())
    if (buf.byteLength > maxBytes) {
      throw new ValidationError(
        `Download exceeds the maximum size of ${maxBytes} bytes`,
      )
    }
    return buf
  }

  const reader = res.body.getReader()
  const chunks: Uint8Array[] = []
  let total = 0

  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      if (!value) continue

      total += value.byteLength
      if (total > maxBytes) {
        // Stop pulling bytes from the sandbox.
        await reader.cancel()
        throw new ValidationError(
          `Download exceeds the maximum size of ${maxBytes} bytes`,
        )
      }
      chunks.push(value)
    }
  } finally {
    reader.releaseLock()
  }

  const out = new Uint8Array(total)
  let offset = 0
  for (const chunk of chunks) {
    out.set(chunk, offset)
    offset += chunk.byteLength
  }
  return out
}

/**
 * Download raw bytes from a URL. Used for file download from the data plane.
 *
 * Retries on transient failures (GET is idempotent).
 *
 * The response body is read with a byte cap (`maxBytes`, default 2 GiB) to
 * protect against a hostile data plane returning an unbounded response.
 */
export async function downloadBytes(opts: {
  url: string
  headers: Record<string, string>
  timeoutMs?: number
  signal?: AbortSignal
  maxBytes?: number
}): Promise<Uint8Array> {
  const {
    url,
    headers,
    timeoutMs = DEFAULT_TIMEOUT_MS,
    signal: userSignal,
    maxBytes = DEFAULT_MAX_DOWNLOAD_BYTES,
  } = opts

  const mergedHeaders: Record<string, string> = {
    "User-Agent": USER_AGENT,
    ...headers,
  }

  try {
    const res = await retryableFetch(
      url,
      { method: "GET", headers: mergedHeaders },
      { timeoutMs, retryable: true, userSignal },
    )

    if (!res.ok) {
      const errorBody = await readErrorBody(res)
      throw mapApiError(res.status, errorBody)
    }

    return await readBodyWithLimit(res, maxBytes)
  } catch (err) {
    if (err instanceof SandboxError) throw err
    if (err instanceof DOMException && err.name === "AbortError") {
      throw new SandboxError("Download aborted", undefined, undefined, {
        cause: err,
      })
    }
    throw new SandboxError(
      `Download error: ${err instanceof Error ? err.message : String(err)}`,
      undefined,
      undefined,
      { cause: err },
    )
  }
}

/**
 * Consume an SSE stream from the exec/stream endpoint.
 *
 * Calls `onEvent` for each parsed event. Returns when the stream ends
 * (server sends `finished: true` or closes the connection).
 *
 * The timeout is an idle timeout — it resets each time a chunk is received,
 * so long-running commands won't spuriously abort.
 *
 * Not retried — POST is not idempotent.
 */
export async function streamSSE<TEvent = ApiExecStreamEvent>(opts: {
  url: string
  headers: Record<string, string>
  body?: unknown
  method?: "GET" | "POST"
  timeoutMs?: number
  signal?: AbortSignal
  onEvent: (event: TEvent) => void
}): Promise<void> {
  const {
    url,
    headers,
    body,
    method = "POST",
    timeoutMs = 300_000,
    signal: userSignal,
    onEvent,
  } = opts
  const controller = new AbortController()
  let timedOut = false
  let timer: ReturnType<typeof setTimeout> | null = setTimeout(() => {
    timedOut = true
    controller.abort()
  }, timeoutMs)

  const signal = composeSignals(controller.signal, userSignal)

  try {
    const init: RequestInit = {
      method,
      headers: {
        "User-Agent": USER_AGENT,
        ...(method === "POST" ? { "Content-Type": "application/json" } : {}),
        ...headers,
      },
      signal,
    }
    if (method === "POST") {
      init.body = JSON.stringify(body ?? {})
    }

    const res = await fetch(url, init)

    if (!res.ok) {
      const errorBody = await readErrorBody(res)
      throw mapApiError(res.status, errorBody)
    }

    if (!res.body) {
      throw new SandboxError("Expected streaming response but got empty body")
    }

    const reader = res.body.getReader()
    const decoder = new TextDecoder()
    let buffer = ""

    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      // Reset the idle timer on each chunk received
      if (timer) {
        clearTimeout(timer)
        timer = setTimeout(() => {
          timedOut = true
          controller.abort()
        }, timeoutMs)
      }

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split("\n")
      buffer = lines.pop() ?? ""

      for (const line of lines) {
        // SSE spec: `data:` with optional leading space on the value.
        if (!line.startsWith("data:")) continue
        const json = line.slice(5).trim()
        if (!json || json === "[DONE]") continue
        try {
          const event = JSON.parse(json) as TEvent
          onEvent(event)
        } catch {
          // Skip malformed events
        }
      }
    }
  } catch (err) {
    if (err instanceof SandboxError) throw err
    if (err instanceof DOMException && err.name === "AbortError") {
      if (timedOut)
        throw new TimeoutError(`Stream timed out after ${timeoutMs}ms`)
      throw new SandboxError("Stream aborted", undefined, undefined, {
        cause: err,
      })
    }
    throw new SandboxError(
      `Stream error: ${err instanceof Error ? err.message : String(err)}`,
      undefined,
      undefined,
      { cause: err },
    )
  } finally {
    if (timer) clearTimeout(timer)
  }
}

File: packages/ruby-sdk/lib/superserve/http.rb
require "faraday"
require "json"
require "time"

module Superserve
  # HTTP client wrapping Faraday with retry on idempotent methods,
  # exponential backoff with jitter, and SSE stream parsing.
  module HTTP
    DEFAULT_TIMEOUT = 30
    DEFAULT_STREAM_TIMEOUT = 300
    USER_AGENT = "superserve-ruby/#{Superserve::VERSION} (ruby/#{RUBY_VERSION})".freeze

    MAX_ATTEMPTS = 3
    BASE_BACKOFF = 0.1
    MAX_BACKOFF = 30.0
    RETRY_STATUS_CODES = [429, 502, 503, 504].freeze
    IDEMPOTENT_METHODS = [:get, :delete].freeze

    # Build a reusable Faraday connection. Pass a single connection per Sandbox
    # for connection pooling (analogous to httpx.Client in the Python SDK).
    def self.build_connection(timeout: DEFAULT_TIMEOUT)
      Faraday.new do |conn|
        conn.options.timeout = timeout
        conn.options.open_timeout = timeout
        conn.options.read_timeout = timeout
      end
    end

    # Make a JSON API request and return the parsed response body.
    # Returns nil for 204 No Content. Raises typed SandboxError on non-2xx.
    # Retries GET/DELETE on transient failures (429, 502/503/504, network errors).
    def self.api_request(method, url, headers:, json_body: nil, timeout: DEFAULT_TIMEOUT, connection: nil)
      conn = connection || build_connection(timeout: timeout)
      method_sym = method.to_sym
      body = json_body.nil? ? nil : JSON.generate(json_body)
      merged_headers = default_headers(headers, content_type: "application/json")

      response = do_request_with_retry(conn, method_sym, url, body, merged_headers, timeout)

      return nil if response.status == 204

      unless success?(response.status)
        raise Superserve.map_api_error(response.status, parse_error_body(response.body))
      end

      response.body.to_s.empty? ? nil : JSON.parse(response.body)
    end

    # Upload raw bytes (POST). Never retried — POST is not idempotent.
    def self.upload_bytes(url, headers:, content:, timeout: DEFAULT_TIMEOUT, connection: nil)
      conn = connection || build_connection(timeout: timeout)
      merged_headers = default_headers(headers, content_type: "application/octet-stream")

      begin
        response = conn.run_request(:post, url, content, merged_headers) do |req|
          req.options.timeout = timeout
          req.options.read_timeout = timeout
        end
      rescue Faraday::TimeoutError
        raise Superserve::TimeoutError.new("Upload timed out after #{timeout}s")
      rescue Faraday::ConnectionFailed => e
        raise Superserve::SandboxError.new("Upload error: #{e.message}")
      end

      unless success?(response.status)
        raise Superserve.map_api_error(response.status, parse_error_body(response.body))
      end

      nil
    end

    # Download raw bytes (GET). Retries on transient failures.
    def self.download_bytes(url, headers:, timeout: DEFAULT_TIMEOUT, connection: nil)
      conn = connection || build_connection(timeout: timeout)
      merged_headers = default_headers(headers)

      response = do_request_with_retry(conn, :get, url, nil, merged_headers, timeout)

      unless success?(response.status)
        raise Superserve.map_api_error(response.status, parse_error_body(response.body))
      end

      response.body.to_s
    end

    # Consume an SSE stream. Calls the block for each parsed event hash.
    # Idle timeout (Net::HTTP read_timeout) resets per chunk under the hood.
    # Supports POST (with body) and GET. Never retried — non-idempotent semantics.
    def self.stream_sse(url, headers:, json_body: nil, method: :post, timeout: DEFAULT_STREAM_TIMEOUT, connection: nil, &on_event)
      raise ArgumentError, "stream_sse requires a block" unless on_event

      conn = connection || build_connection(timeout: timeout)
      method_sym = method.to_sym
      body = json_body.nil? ? nil : JSON.generate(json_body)
      merged_headers = default_headers(
        headers,
        content_type: method_sym == :post ? "application/json" : nil
      )

      buffer = String.new

      begin
        response = conn.run_request(method_sym, url, body, merged_headers) do |req|
          req.options.timeout = timeout
          req.options.read_timeout = timeout
          req.options.on_data = proc do |chunk, _bytes|
            buffer << chunk.to_s
            loop do
              newline_idx = buffer.index("\n")
              break if newline_idx.nil?

              line = buffer.slice!(0..newline_idx).chomp

              next unless line.start_with?("data:")

              data = line[5..].to_s.strip
              next if data.empty? || data == "[DONE]"

              begin
                event = JSON.parse(data)
                on_event.call(event)
              rescue JSON::ParserError
                # Skip malformed events
              end
            end
          end
        end
      rescue Faraday::TimeoutError
        raise Superserve::TimeoutError.new("Stream timed out after #{timeout}s")
      rescue Faraday::ConnectionFailed => e
        raise Superserve::SandboxError.new("Stream error: #{e.message}")
      end

      # On error, `buffer` retains any unparsed content (typically the full
      # error JSON body, since error responses don't use `data:` framing).
      unless success?(response.status)
        raise Superserve.map_api_error(response.status, parse_error_body(buffer))
      end

      nil
    end

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    # @api private
    def self.do_request_with_retry(connection, method, url, body, headers, timeout)
      retryable = IDEMPOTENT_METHODS.include?(method)
      max_attempts = retryable ? MAX_ATTEMPTS : 1
      attempt = 1

      loop do
        begin
          response = connection.run_request(method, url, body, headers) do |req|
            req.options.timeout = timeout
            req.options.read_timeout = timeout
          end
        rescue Faraday::TimeoutError
          raise Superserve::TimeoutError.new("Request timed out after #{timeout}s")
        rescue Faraday::ConnectionFailed => e
          if !retryable || attempt >= max_attempts
            raise Superserve::SandboxError.new("Network error: #{e.message}")
          end
          sleep(compute_backoff(attempt))
          attempt += 1
          next
        end

        if retryable && RETRY_STATUS_CODES.include?(response.status) && attempt < max_attempts
          delay = nil
          if response.status == 429
            delay = parse_retry_after(response.headers["retry-after"] || response.headers["Retry-After"])
          end
          delay ||= compute_backoff(attempt)
          sleep(delay) if delay > 0
          attempt += 1
          next
        end

        return response
      end
    end

    # Exponential backoff with ±20% jitter, capped at MAX_BACKOFF.
    # Attempt 1 → ~100ms, 2 → ~200ms, 3 → ~400ms.
    # @api private
    def self.compute_backoff(attempt)
      base = BASE_BACKOFF * (2**(attempt - 1))
      jitter = base * (0.8 + rand * 0.4)
      [jitter, MAX_BACKOFF].min
    end

    # Parse a Retry-After header (numeric seconds or HTTP-date).
    # Returns seconds capped at MAX_BACKOFF, or nil if invalid.
    # @api private
    def self.parse_retry_after(value)
      return nil if value.nil?
      trimmed = value.to_s.strip
      return nil if trimmed.empty?

      begin
        seconds = Float(trimmed)
        return [[seconds, 0.0].max, MAX_BACKOFF].min
      rescue ArgumentError
      end

      begin
        dt = Time.httpdate(trimmed)
        delta = dt - Time.now
        return [[delta, 0.0].max, MAX_BACKOFF].min
      rescue ArgumentError
        nil
      end
    end

    # @api private
    def self.default_headers(extra, content_type: nil)
      h = { "User-Agent" => USER_AGENT }
      h["Content-Type"] = content_type if content_type
      h.merge(extra)
    end

    # @api private
    def self.parse_error_body(body)
      return {} if body.nil? || body.to_s.empty?
      parsed = JSON.parse(body.to_s)
      parsed.is_a?(Hash) ? parsed : {}
    rescue JSON::ParserError
      truncated = body.to_s[0, 500]
      { "error" => { "message" => truncated.empty? ? nil : truncated } }
    end

    # @api private
    def self.success?(status)
      status >= 200 && status < 300
    end
  end
end
